File: homekit_server/kmarket.py
import json
import re
import datetime
import time

import requests
from six import string_types

from homekit_server.utils import get_cache, format_mtime


def get_kmarket_data(store_id):
    """
    Get K-Market data for the given Store ID.

    Store IDs are (empirically) of the form `Knnn` where n is a number.

    :param store_id: The store ID to return data for.
    :return: dict
    """
    assert isinstance(store_id, string_types)
    cache_key = "kmarket-%s" % store_id
    cache = get_cache()
    data = cache.get(cache_key)
    date = datetime.date.today().isoformat()
    if data and data["$date"] == date:  # Fresh cached data
        data["$cached"] = True
        return data
    data = _get_live_kmarket_data(store_id)
    data["$date"] = date
    data["$mtime"] = format_mtime()
    cache.set(cache_key, data, timeout=300)
    return data


def _get_live_kmarket_data(store_id):
    cb = "u%d" % time.time()
    resp = requests.get(
        url="http://www.k-ruoka.fi/api/usertools/GetUserStoreSpecifics?callback=%s" % cb,
        headers={
            "Cookie": "UserStores=UserStores=%s" % store_id,
            "Referer": "http://www.k-ruoka.fi/",
        }
    )
    match = re.match("^%s\((\[.+\])\);" % cb, resp.text)
    return json.loads(match.group(1))[0]

File: homekitd/kmarket.js
const {cached} = require('./cache');
const fetch = require('node-fetch');
const R = require('ramda');
const uid = require('./uid');
const moment = require('moment');

const getDateStamp = () => moment().format('YYYY-MM-DD');

function downloadData(storeId) {
  const cb = uid('u');
  return fetch(
    'http://www.k-ruoka.fi/api/usertools/GetUserStoreSpecifics?callback=' + cb,
    {
      headers: {
        'Cookie': `UserStores=UserStores=${storeId}`,
        'Referer': 'http://www.k-ruoka.fi/'
      }
    }
  )
    .then((res) => res.text())
    .then((text) => {
      const re = new RegExp('^' + cb + '\\((\\[.+\\])\\);');
      const m = re.exec(text);
      if (!m) {
        return Promise.reject('Unable to parse K-Market response');
      }
      return JSON.parse(m[1])[0];
    })
    .then(R.merge({
      $date: getDateStamp()
    }));
}

module.exports = (storeId) => {
  return cached(
    `kmarket-${storeId}`,
    () => downloadData(storeId),
    60 * 10,
    (val) => (val && val.$date == getDateStamp())
  );
};
